src/protocol/ContactRequestChannel.cpp

```diff
@@ -0,0 +1,305 @@
+            c.category() == QChar::Other_Control ||
+            c.isNonCharacter())
+            return false;
+    }
+
+    return true;
+}
+
+QString ContactRequestChannel::nickname() const
+{
+    return m_nickname;
+}
+
+void ContactRequestChannel::setNickname(const QString &nickname)
+{
+    if (direction() != Outbound) {
+        BUG() << "Request messages can only be set on outbound messages";
+        return;
+    }
+
+    if (isOpened() || identifier() >= 0) {
+        BUG() << "Request data must be set before opening channel";
+        return;
+    }
+
+    if (!isAcceptableNickname(nickname)) {
+        BUG() << "Outbound contact request nickname isn't acceptable:" << nickname;
+        return;
+    }
+
+    m_nickname = nickname;
+}
+
+bool ContactRequestChannel::allowInboundChannelRequest(const Data::Control::OpenChannel *request, Data::Control::ChannelResult *result)
+{
+    using namespace Data::ContactRequest;
+
+    // If this connection is already KnownContact, report that the request is accepted
+    if (connection()->purpose() == Connection::Purpose::KnownContact) {
+        QScopedPointer<Response> response(new Response);
+        response->set_status(Response::Accepted);
+        result->SetAllocatedExtension(Data::ContactRequest::response, response.take());
+        return false;
+    }
+
+    // We'll only accept requests on inbound connections with an unknown purpose
+    if (connection()->direction() != Connection::ServerSide ||
+        connection()->purpose() != Connection::Purpose::Unknown)
+    {
+        result->set_common_error(Data::Control::ChannelResult::UnauthorizedError);
+        result->set_error_message(QStringLiteral("Only a new client may use this channel").toStdString());
+        return false;
+    }
+
+    // Only allow one ContactRequestChannel
+    if (connection()->findChannel<ContactRequestChannel>()) {
+        result->set_common_error(Data::Control::ChannelResult::UnauthorizedError);
+        result->set_error_message(QStringLiteral("Only one instance of this channel may be created").toStdString());
+        return false;
+    }
+
+    // Require HiddenServiceAuth
+    if (!connection()->hasAuthenticated(Connection::HiddenServiceAuth)) {
+        result->set_common_error(Data::Control::ChannelResult::UnauthorizedError);
+        result->set_error_message(QStringLiteral("Only authenticated clients may use this channel").toStdString());
+        return false;
+    }
+
+    if (!request->HasExtension(Data::ContactRequest::contact_request)) {
+        result->set_error_message(QStringLiteral("Expected a request object").toStdString());
+        return false;
+    }
+
+    ContactRequest contactData = request->GetExtension(Data::ContactRequest::contact_request);
+    QString nickname = QString::fromStdString(contactData.nickname());
+    QString message = QString::fromStdString(contactData.message_text());
+
+    if (message.size() > Data::ContactRequest::MessageMaxCharacters ||
```

Comment:
It's not. This says that the message is invalid if there are more than MessageMaxCharacters. So the maximum valid number of characters is the value of MessageMaxCharacters. That fits with my expectation from the word 'max'.